### PR TITLE
drivers: ethernet: introduce SMSC91C111 driver

### PR DESCRIPTION
**Pull request by SgrrZhf**
This PR has 4 parts  
 1. Introduce SMSC91C111 ethernet driver.
 2. Enable SMSC91C111 on the `fvp_base_revc_2xaemv8a` and `fvp_baser_aemv8r` boards.
 3. Enable 2 network samples below to demonstrate how to use it.
     - dhcpv4_client
     - zperf
 5. Add some documents to descript it.

### CODE REVIEW

File: samples/net/zperf/src/main.c [hunk 1]
```diff
@@ -14,6 +14,53 @@
...
+
+	for (i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+		char buf[NET_IPV4_ADDR_LEN];
+
+		if (iface->config.ip.ipv4->unicast[i].addr_type !=
+							NET_ADDR_DHCP) {
+			continue;
+		}
+
+		LOG_INF("Your address: %s",
```
Comment (povergoing) [inline samples/net/zperf/src/main.c:46]: Would you please fix the alignment of the following lines?

File: drivers/ethernet/eth_smsc91x.c [hunk 2]
```diff
@@ -0,0 +1,850 @@
...
+
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT smsc_lan91c111_mdio
+
+struct mdio_smsc_config {
+	const struct device *eth_dev;
+};
+
+static void mdio_smsc_bus_disable(const struct device *dev)
+{
```
Comment (povergoing) [inline drivers/ethernet/eth_smsc91x.c:798]: ARG_UNUSED(dev) is needed if dev is unused.

File: drivers/ethernet/eth_smsc91x.c [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+}
+
+static void mdio_smsc_bus_enable(const struct device *dev)
+{
```
Comment (povergoing) [inline drivers/ethernet/eth_smsc91x.c:802]: ditto

File: drivers/ethernet/eth_smsc91x.c [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+	.read = mdio_smsc_read,
+	.write = mdio_smsc_write,
+};
+
+static int mdio_smsc_init(const struct device *dev)
+{
+	/*
+	 * It doesn't need to initialize anything because smsc_init has
+	 * already done it.
+	 */
```
Comment (povergoing) [inline drivers/ethernet/eth_smsc91x.c:839]: ditto, ARG_UNUSED(dev)

File: drivers/ethernet/eth_smsc91x.c [hunk 5]
```diff
@@ -0,0 +1,850 @@
...
+}
+
+static int mdio_smsc_read(const struct device *dev, uint8_t prtad, uint8_t devad, uint16_t *data)
+{
+	const struct mdio_smsc_config *cfg = dev->config;
+	const struct device *eth_dev = cfg->eth_dev;
+	struct eth_context *eth_data = eth_dev->data;
+	struct smsc_data *sc = &eth_data->sc;
+
+	*data = smsc_miibus_readreg(sc, prtad, devad);
```
Comment (povergoing) [inline drivers/ethernet/eth_smsc91x.c:812]: might be better to insert a blank line before `return 0;`

File: drivers/ethernet/eth_smsc91x.c [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+}
+
+static int mdio_smsc_write(const struct device *dev, uint8_t prtad, uint8_t devad, uint16_t data)
+{
+	const struct mdio_smsc_config *cfg = dev->config;
+	const struct device *eth_dev = cfg->eth_dev;
+	struct eth_context *eth_data = eth_dev->data;
+	struct smsc_data *sc = &eth_data->sc;
+
+	smsc_miibus_writereg(sc, prtad, devad, data);
```
Comment (povergoing) [inline drivers/ethernet/eth_smsc91x.c:823]: ditto, insert a blank line

File: drivers/ethernet/eth_smsc91x.c (hunk 4 above)
Comment (SgrrZhf) [reply drivers/ethernet/eth_smsc91x.c:845]: Done

File: samples/net/zperf/src/main.c [hunk 7]
```diff
@@ -29,4 +69,15 @@ void main(void)
...
+
+#ifdef CONFIG_NET_DHCPV4
+	struct net_if *iface;
+
+	LOG_INF("Run dhcpv4 client");
+	net_mgmt_init_event_callback(&mgmt_cb, handler, NET_EVENT_IPV4_ADDR_ADD);
+	net_mgmt_add_event_callback(&mgmt_cb);
+
+	iface = net_if_get_default();
+	net_dhcpv4_start(iface);
```
Comment (rlubos) [inline samples/net/zperf/src/main.c:81]: Generally this commit should not be needed. The sample uses NET_CONFIG lib, which triggers DHCPv4, if enabled. I had no problem obtaining an IPv4 address over DHCPv4 with other boards.

**Review by nordicjm — COMMENTED** [review]
Build system parts look OK, one comment.

File: samples/net/dhcpv4_client/README.rst [hunk 8]
```diff
@@ -145,3 +145,41 @@ an ip address by typing:
 .. code-block:: console
 
     $ ping -I eth1 192.0.2.10
+
+
+Arm FVP
```
[the remark this answers is not in the document]
Comment (SgrrZhf) [reply samples/net/dhcpv4_client/README.rst:150]: If users want to run this sample on `qemu_x86` or `FRDM_K64F`, they need to do some extra network configurations on their host operating system, for example, add a `tap` device.
I want to point out that the Arm FVP board can run this sample directly and users don't need to do other extra network configurations on their host operating system. So I add documentation here.

File: samples/net/zperf/src/main.c (hunk 7 above)
Comment (SgrrZhf) [reply samples/net/zperf/src/main.c:81]: Thank you for reminding me, I'll try it and remove this commit.

**Review by povergoing — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Pull request merged**